test/hotspot/jtreg/serviceability/jvmti/ObjectMonitorUsage/ObjectMonitorUsage.java

```diff
@@ -0,0 +1,326 @@
+    final static int NUMBER_OF_ENTERING_THREADS = 4;
+    final static int NUMBER_OF_WAITING_THREADS  = 4;
+    final static int NUMBER_OF_THREADS = NUMBER_OF_ENTERING_THREADS + NUMBER_OF_WAITING_THREADS;
+
+    static Object lockCheck = new Object();
+
+    native static int getRes();
+    native static int waitsToEnter();
+    native static int setTestedMonitor(Object monitor);
+    native static void check(Object obj, Thread owner,
+                             int entryCount, int waiterCount, int notifyWaiterCount);
+
+    static void log(String msg) {
+        System.out.println(msg);
+    }
+
+    static String vtag(boolean isVirtual) {
+        return isVirtual ? "virtual" : "platform";
+    }
+
+    static void sleep(long millis) {
+        try {
+            Thread.sleep(millis);
+        } catch (InterruptedException e) {
+            // ignore
+        }
+    }
+
+    static Thread startTask(int idx, TestTask task, boolean isVirtual, String kind) {
+        Thread thread = isVirtual ? Thread.ofVirtual().name(kind + "VT" + idx).start(task)
+                                  : Thread.ofPlatform().name(kind + "PT" + idx).start(task);
+        task.waitReady();
+        return thread;
+    }
+
+    static Thread[] startWaitingThreads(boolean isVirtual) {
+        Thread[] threads = new Thread[NUMBER_OF_WAITING_THREADS];
+        for (int i = 0; i < NUMBER_OF_WAITING_THREADS; i++) {
+            // the WaitingTask has to wait to be notified in lockCheck.wait()
+            threads[i] = startTask(i, new WaitingTask(), isVirtual, "Waiting");
+        }
+        return threads;
+    }
+
+    static Thread[] startEnteringThreads(boolean isVirtual) {
+        Thread[] threads = new Thread[NUMBER_OF_ENTERING_THREADS];
+        for (int i = 0; i < NUMBER_OF_ENTERING_THREADS; i++) {
+            // the EnteringTask has to be blocked at the lockCheck enter
+            threads[i] = startTask(i, new EnteringTask(), isVirtual, "Entering");
+        }
+        while (waitsToEnter() < NUMBER_OF_ENTERING_THREADS) {
+            sleep(1);
+        }
+        return threads;
+    }
+
+    static void joinThreads(Thread[] threads) {
+        try {
+            for (Thread t : threads) {
+                t.join();
+            }
+        } catch (InterruptedException e) {
+            throw new Error("Unexpected " + e);
+        }
+    }
+
+    /* Scenario #0:
+     * - owning:         0
+     * - entering:       0
+     * - re-entering:    0
+     * - to be notified: N
+     */
+    static void test0(boolean isVirtual) {
+        String vtag = vtag(isVirtual);
+        log("\n###test0: started " + vtag);
+
+        Thread[] wThreads = startWaitingThreads(isVirtual);
+
+        // entry count: 0
+        // count of threads waiting to enter:       0
+        // count of threads waiting to re-enter:    0
+        // count of threads waiting to be notified: NUMBER_OF_WAITING_THREADS
+        check(lockCheck, null, 0, // no owner thread
```

Comment:
AFAICS you are racing here, as the waiting threads can set `ready` but not yet have called `wait()`. To know for sure they are in `wait()` you need to acquire the monitor before checking (and release it again so that you test the no-owner scenario as intended). But this should probably be done inside `startWaitingThreads`.

Reply:
Thank you for the comment. Will implement similar approach as in the `startEnteringThreads()`.